sky/spot/core.py

```diff
@@ -0,0 +1,299 @@
+                          'Restarting controller for latest status...'
+                          f'{colorama.Style.RESET_ALL}')
+
+        rich_utils.force_update_status('[cyan] Checking spot jobs - restarting '
+                                       'controller[/]')
+        handle = sky.start(spot_utils.SPOT_CONTROLLER_NAME)
+        controller_status = status_lib.ClusterStatus.UP
+        rich_utils.force_update_status('[cyan] Checking spot jobs[/]')
+
+    assert handle is not None, (controller_status, refresh)
+
+    backend = backend_utils.get_backend_from_handle(handle)
+    assert isinstance(backend, backends.CloudVmRayBackend)
+
+    code = spot_utils.SpotCodeGen.get_job_table()
+    returncode, job_table_payload, stderr = backend.run_on_head(
+        handle,
+        code,
+        require_outputs=True,
+        stream_logs=False,
+        separate_stderr=True)
+
+    try:
+        subprocess_utils.handle_returncode(returncode,
+                                           code,
+                                           'Failed to fetch managed spot jobs',
+                                           job_table_payload + stderr,
+                                           stream_logs=False)
+    except exceptions.CommandError as e:
+        raise RuntimeError(str(e)) from e
+
+    jobs = spot_utils.load_spot_job_queue(job_table_payload)
+    if skip_finished:
+        # Filter out the finished jobs. If a multi-task job is partially
+        # finished, we will include all its tasks.
+        non_finished_tasks = list(
+            filter(lambda job: not job['status'].is_terminal(), jobs))
+        non_finished_job_ids = {job['job_id'] for job in non_finished_tasks}
+        jobs = list(
+            filter(lambda job: job['job_id'] in non_finished_job_ids, jobs))
+    return jobs
+
+
+@usage_lib.entrypoint
+# pylint: disable=redefined-builtin
+def cancel(name: Optional[str] = None,
+           job_ids: Optional[List[int]] = None,
+           all: bool = False) -> None:
+    # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
+    """Cancel managed spot jobs.
+
+    Please refer to the sky.cli.spot_cancel for the document.
+
+    Raises:
+        sky.exceptions.ClusterNotUpError: the spot controller is not up.
+        RuntimeError: failed to cancel the job.
+    """
+    job_ids = [] if job_ids is None else job_ids
+    handle = backend_utils.is_controller_accessible(
+        controller_type=controller_utils.Controllers.SPOT_CONTROLLER,
+        stopped_message='All managed spot jobs should have finished.')
+
+    job_id_str = ','.join(map(str, job_ids))
+    if sum([len(job_ids) > 0, name is not None, all]) != 1:
+        argument_str = f'job_ids={job_id_str}' if len(job_ids) > 0 else ''
+        argument_str += f' name={name}' if name is not None else ''
+        argument_str += ' all' if all else ''
+        raise ValueError('Can only specify one of JOB_IDS or name or all. '
```

Comment:
do we need `with ux_utils.print_exception_no_traceback()`? ditto for all raises in this file